addons/addon-base-raas/packages/base-raas-services/lib/data-source/access-strategy/roles-only/__tests__/filesystem-role-service.test.js

```diff
@@ -0,0 +1,363 @@
+const FilesystemRoleService = require('../filesystem-role-service');
+
+const createStudy = ({
+  id = 'study-1',
+  category = 'Organization',
+  accountId = '1122334455',
+  awsPartition = 'aws',
+  bucketAccess = 'roles',
+  bucket = 'bucket-1',
+  qualifier = 'swb-IhsKhN8GsLneiis11ujlb8',
+  appRoleArn = 'arn:aws:iam::684277579687:role/swb-IhsKhN8GsLneiis11ujlb8-app-1607537845811',
+  accessType = 'readwrite',
+  envPermission = { read: true, write: true },
+  folder = '/',
+  kmsScope = 'none',
+} = {}) => ({
+  id,
+  category,
+  accountId,
+  awsPartition,
+  bucketAccess,
+  bucket,
+  qualifier,
+  appRoleArn,
+  accessType,
+  envPermission,
+  folder,
+  kmsScope,
+});
+
+const createAppRole = ({
+  arn = 'arn:aws:iam::684277579687:role/swb-IhsKhN8GsLneiis11ujlb8-app-1607537845811',
+  accountId = '1122334455',
+  mainRegion = 'us-east-1',
+  awsPartition = 'aws',
+  bucket = 'bucket-1',
+  bucketRegion = 'us-east-1',
+  status = 'pending',
+  name = 'swb-IhsKhN8GsLneiis11ujlb8-app-1607537845811',
+  qualifier = 'swb-IhsKhN8GsLneiis11ujlb8',
+  boundaryPolicyArn = 'arn:aws:iam::684277579687:policy/swb-IhsKhN8GsLneiis11ujlb8-app-1607537845811',
+  studies = {
+    'study-1': {
+      accessType: 'readonly',
+      kmsScope: 'none',
+      folder: '/',
+    },
+  },
+} = {}) => ({
+  accountId,
+  arn,
+  name,
+  mainRegion,
+  awsPartition,
+  bucket,
+  bucketRegion,
+  status,
+  qualifier,
+  boundaryPolicyArn,
+  studies,
+});
+
+const createAdminContext = ({ uid = 'uid-admin' } = {}) => ({
+  principalIdentifier: { uid },
+  principal: { isAdmin: true, userRole: 'admin', status: 'active' },
+});
+
+describe('DataSourceBucketService', () => {
+  let container;
+  let service;
+  let appRoleService;
+  let usageService;
+
+  beforeEach(async () => {
+    // Initialize services container and register dependencies
+    container = new ServicesContainer();
+
+    container.register('dbService', new DbService());
+    container.register('jsonSchemaValidationService', new JsonSchemaValidationService());
+    container.register('settings', new SettingsService());
+    container.register('aws', new Aws());
+    container.register('log', new Logger());
+    container.register('pluginRegistryService', new PluginRegistryService());
+    container.register('authorizationService', new AuthService());
+    container.register('auditWriterService', new AuditService());
+    container.register('roles-only/applicationRoleService', new ApplicationRoleService());
+    container.register('resourceUsageService', new ResourceUsageService());
+    container.register('roles-only/filesystemRoleService', new FilesystemRoleService());
+    await container.initServices();
+
+    service = await container.find('roles-only/filesystemRoleService');
+    appRoleService = await container.find('roles-only/applicationRoleService');
+    usageService = await container.find('resourceUsageService');
+  });
+
+  describe('when allocating a role', () => {
```

Comment:
Can we add a few test cases are following cases:

1. We found the appropriate filesystem role but couldn't add study, env, account via resourceUsageService because of network failure or something. Calling the allocateRole method should work successfully without side effects.

2. The above case when we didn't find a filesystem role and created a new one. But couldn't add the (study, env, account) via resourceUsageService

3. The account doesn't exist in fileSystemRole, so we update the policy but method fails while saving the entity so we can't call resourceUsageService as well. A retry on the method should ideally work.

Reply:
Great suggestions. I am going to write more unit tests and document the different error scenarios that can happen and how to recover from them.  I am thinking of doing that once we have the whole solution in a working state.